### PR TITLE
Rpc: Add getConfirmedSignaturesForAddress

### LINKED ISSUE #9317: Add JSON RPC primitives for fetching the list of transactions that affect a given address
For the history that an API node maintains, it's not possible to fetch the list of transactions that affect a given address.   This is a feature that is starting to be widely requested.

These two new JSON RPC methods could be path to enabling this feature:

1. `getConfirmTransactionsForAddress`.   Input an address, return all the transaction signatures (paginated by slot?).   For this we'll require a new rocksdb column to store the mapping of Address -> (Signature, Slot)

2. `getConfirmedTransactionBySignature`.  Input a signature, return the transaction.  This could be build on top of Blockstore's `get_transaction_status()`

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
For the history that an API node maintains, it's not possible to fetch the list of transactions that affect a given address. Now that we index signatures by address, the next step would be to return those signatures 

#### Summary of Changes
- Fix AddressSignatures blockstore column to handle multiple transactions per address-slot (also clean up some iterator indexes in tests so changes are on in one place in the future)
- Add blockstore apis to return signatures by address (forward-looking to returning full transactions by address)
- Add `getConfirmedSignaturesforAddress` rpc endpoint to expose signatures-by-address data

Fixes #9317

### CODE REVIEW

File: core/src/rpc.rs [hunk 1]
```diff
@@ -514,6 +514,22 @@ impl JsonRpcRequestProcessor {
             Ok(None)
         }
     }
+
+    pub fn get_address_confirmed_signatures(
+        &self,
+        pubkey: Pubkey,
+        start_slot: Option<Slot>,
+        end_slot: Option<Slot>,
```
Comment (mvines) [inline core/src/rpc.rs:522]: I think we should make start/end slot required, and cap the difference between them to some limit.  Maybe 10,000 slots (~1 hour).

App devs are lazy and they'll just not provide any start/end slot and the node'll get crushed with all the `(0, std::u64::MAX)` queries

Comment (CriesofCarrots) [reply core/src/rpc.rs:522]: Fair enough

**Comment by CriesofCarrots** [comment]
@mvines , ready for another look.
Currently, both start_slot and end_slot are required, although I did wonder about making it consistent with `getConfirmedBlocks`, which only requires start_slot. Do you have any opinions on that?

**Comment by mvines** [comment]
> @mvines , ready for another look.
> Currently, both start_slot and end_slot are required, although I did wonder about making it consistent with `getConfirmedBlocks`, which only requires start_slot. Do you have any opinions on that?

Only requiring `start_slot` seems fine, but then we'd want to return the range that was searched in the response so they know where to search next.  Maybe we should return the range regardless.

**Comment by CriesofCarrots** [comment]
> > @mvines , ready for another look.
> Only requiring `start_slot` seems fine, but then we'd want to return the range that was searched in the response so they know where to search next. Maybe we should return the range regardless.

The optional `end_slot` is more problematic with `getConfirmedTransactionsForAddress`, which I'm working on now. It needs an optional `encoding` parameter as well, and the jsonrpc crate doesn't handle multiple optional positional parameters very well.

So I'd like to leave `end_slot` as-is. In that case, I don't think we should return the range. Sound okay?

File: docs/src/apps/jsonrpc-api.md [hunk 2]
```diff
@@ -346,6 +347,31 @@ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0","id":1,"m
...
+#### Parameters:
+
+* `<string>` - account address as base-58 encoded string
+* `<u64>` - start slot, inclusive
+* `<u64>` - end slot, inclusive
+
+#### Results:
+
+The result field will be an array of:
+* `<string>` - transaction signature as base-58 encoded string
```
Comment (mvines) [inline docs/src/apps/jsonrpc-api.md:363]: nit: let's define an ordering.  The signatures will be ordered based on the slot in which they were confirmed in, from lowest to highest slot

**Pull request merged**